File: php/subspider.php
<?php
    include 'Spider.php';
    class SubSpider extends Spider {
        function get_json($url="", $method ="", $referer="",$params=array()) {
            $pagina = parent::request($url, $method, '', $params);

            $pagina = utf8_encode("$pagina");
            $pagina = preg_replace("<!--[\s\S]*?-->", "", $pagina);
            $pagina = preg_replace("/&ccedil;/", 'c', $pagina);
            $pagina = preg_replace("/&atilde;/", 'a', $pagina);


            $info = array();
            $ultima_secao = "";
            while(true) {
                preg_match('/<table width[\s\S]*?>([\s\S]*?)<\/table>([\s\S]*)/', $pagina, $matches);
                $conteudo = $matches[1];
                $restante = $matches[2];
                preg_match('/<tr>\s*<td class="secao".*?>\s*(.*?)\s*<\/td>\s*<\/tr>/', $conteudo, $secao_match);
                $secao = $secao_match[1];

                if($secao) {
                    print_r($restante);
                    $ultima_secao = $secao;
                    $info[$secao] = $this->extrair_valores ($conteudo);
                }
                else {
                    $lista = $this->extrair_valores($restante);
                    $merge = $info[$ultima_secao] + $lista;
                    $info[$ultima_secao] = $merge;
                    break;
                }
                $pagina = $restante;

            }
            return json_encode($info, JSON_UNESCAPED_UNICODE);
        }
        function extrair_valores ($conteudo) {
            preg_match_all('/<td[\s\S]*?class="titulo"[\s\S]*?>(?:&nbsp;)?([\s\S]*?):(?:&nbsp;)?<\/td>\s*<td .*?class="valor"[\s\S]*?>(?:&nbsp;)?([\s\S]*?)<\/td>/', $conteudo, $valores_match);
            $titulos = $valores_match[1];
            $valores = $valores_match[2];

            for ($i = 0; $i < count($titulos); $i++) {
                $lista[$titulos[$i]] = $valores[$i];
            }
            return array_combine($titulos, $valores);
        }
    }

?>

File: ruby/subspider.rb
require 'net/http'
require 'json'
load "spider.rb"

class SubSpider < Spider
    alias :super_post :post
    alias :super_get :get
    def content_type (response)
        /charset=(?<ct>.+)/ =~ response['content-type']
        return ct
    end
    def valid_result (body)
        (/class="erro"/ =~ body) == nil
    end
    def string_encoded (method, path, params = {})
        if method == :post
            response = super_post path, params
        else
            response = super_get path, params
        end
        body = response.body
        contenttype = content_type(response)
        body = body.force_encoding(contenttype).encode('utf-8')
        #mudando algumas conjuntos de caracteres
        body.gsub!(/&ccedil;/, 'ç')
        body.gsub!(/&atilde;/, 'ã')
        return body
    end
    def json_from (method, path, params = {})
        body = string_encoded method, path, params
        #se a entrada for invalida
        if !valid_result body
            return nil
        end
        info = {}
        last_secao = nil
        loop do
        /<table width.*?>(?<conteudo>.*?)<\/table>(?<restante>.*)/m =~ body
        /<tr>\s*<td class="secao".*?>\s*(?<secao>.*?)\s*<\/td>\s*<\/tr>(?<conteudo>.*)/m =~conteudo
            break if secao == nil
            last_secao = secao
            loop do
                /<td .*?class="titulo".*?>(&nbsp;)?(?<titulo>.*?):(&nbsp;)?<\/td>\s*<td .*?class="valor".*?>(&nbsp;)?(?<valor>.*?)<\/td>(?<conteudo>.*)/m =~ conteudo
                break if titulo == nil
                if info[secao] == nil
                    info[secao] = {titulo => valor}
                else
                    info[secao][titulo] = valor
                end
            end
            body = restante
        end
        /<table width.*?>(?<conteudo>.*?)<\/table>(?<restante>.*)/m =~ body
        /<td .*?class="titulo".*?>(&nbsp;)?(?<titulo>.*?):(&nbsp;)?<\/td>\s*<td .*?class="valor".*?>(&nbsp;)?(?<valor>.*?)<\/td>/m =~ conteudo
        info[last_secao][titulo] = valor
        />\s*?<table width.*?>(?<conteudo>.*?)<\/table>/m =~ restante
        />\s*?<td .*?class="titulo".*?>(&nbsp;)?(?<titulo>.*?):(&nbsp;)?<\/td>\s*<td .*?class="valor".*?>(&nbsp;)?(?<valor>.*?)<\/td>/m =~ conteudo
        info[last_secao][titulo] = valor
        return info.to_json
    end
end
















